### PR TITLE
[WIP] Todo reminder notification

### PR DESCRIPTION
**Pull request by rommsen**
This is a first try to implement a reminder notification for todos. 

I added a send_reminder_notification script which queries the read model for all reminders in the past that have not been handled (reminded column, see migration).

The script uses the command bus to dispatch RemindTodoAssignee commands.
CommandHandler checks with the read model if the reminder is still valid (uses var_dumps for testing purposes).
The todo model sets a "reminded" flag to prevent double notifications.

Questions:
- do you think that this is the right approach 
- where should I put the mail subscriber, I chose: App\Mail\SendTodoReminderMailSubscriber

Todos:
- tests
- write a real mail handler
- cronjob for the worker

### CODE REVIEW

**Comment by prolic** [comment]
First of all, please create PR to develop branch.

File: migrations/Version20160308110616.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+class Version20160308110616 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $todo = $schema->getTable(Table::TODO);
+        $todo->addColumn('reminded', 'boolean', ['default' => 0, 'notnull' => false, 'length' => 1]);
```
Comment (prolic) [inline migrations/Version20160308110616.php]: I'd say, let's use a distinct table for reminders, not the same todo table.

File: src/Model/Todo/Handler/RemindTodoAssigneeHandler.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    public function __invoke(RemindTodoAssignee $command)
+    {
+        $todo = $this->todoList->get($command->todoId());
+        if (!$todo) {
+            throw TodoNotFound::withTodoId($command->todoId());
+        }
+
+        $todoData = $this->todoFinder->findById($command->todoId()->toString());
+
+        if ($todoData->reminded) {
```
Comment (prolic) [inline src/Model/Todo/Handler/RemindTodoAssigneeHandler.php]: I'd say, you need more arguments on the command. You then have to check the command data with the data in the current todo. So when the reminder gets edited, the old command is invalid.

File: src/Model/Todo/Todo.php [hunk 3]
```diff
@@ -156,6 +171,14 @@ public function reminder()
     }
 
     /**
+     * @return bool
+     */
+    public function assigneeWasReminded()
```
Comment (prolic) [inline src/Model/Todo/Todo.php]: Irrelevant. An assignee could also be notified multiple times.

File: src/Model/Todo/Todo.php [hunk 4]
```diff
@@ -231,6 +253,16 @@ protected function whenDeadlineWasAddedToTodo(DeadlineWasAddedToTodo $event)
...
     {
         $this->reminder = $event->reminder();
+        $this->reminded = false;
+    }
+
+    /**
+     * @param TodoAssigneeWasReminded $event
+     * @return void
+     */
+    protected function whenTodoAssigneeWasReminded(TodoAssigneeWasReminded $event)
```
Comment (prolic) [inline src/Model/Todo/Todo.php]: Irrelevant. An assignee could also be notified multiple times.

File: src/Projection/Todo/TodoProjector.php [hunk 5]
```diff
@@ -109,6 +110,23 @@ public function onReminderWasAddedToTodo(ReminderWasAddedToTodo $event)
             Table::TODO,
             [
                 'reminder' => $event->reminder()->toString(),
+                'reminded' => 0,
```
Comment (prolic) [inline src/Projection/Todo/TodoProjector.php]: As I said earlier, use a distinct reminder table.

File: src/Model/Todo/Todo.php (hunk 4 above)
Comment (rommsen) [inline src/Model/Todo/Todo.php]: Yeah but not about the same reminder. This is what is prevented here. The assignee should only be notified again if the reminder changes, therefore I set $this->reminded to false during whenReminderWasAddedToTodo

File: src/Model/Todo/Handler/RemindTodoAssigneeHandler.php (hunk 2 above)
Comment (rommsen) [inline src/Model/Todo/Handler/RemindTodoAssigneeHandler.php]: Isnt this already the case already? If the reminder is changed to something in the future the command is dropped (Line 57), if it is changed to something in the past it doesnt matter, as long as the notification is not already send.

File: src/Model/Todo/Todo.php (hunk 4 above)
Comment (prolic) [inline src/Model/Todo/Todo.php]: ok

**Comment by rommsen** [comment]
I am currently refactoring to a distinct reminder read model and more data in the command. I will create another PR to develop branch when finished.

**Pull request closed without merging**